### ISSUE TITLE
networkmanager: slow DHCP offer, interface stays disconnected

### ISSUE THREAD

**Issue by damirsamardzic** [greshake/i3status-rust#991]
I'm working in dev environments with multiple network interfaces where DHCP servers are often responding with delay (dnsmasq with ping check as an example) - sometimes I see up to 5 seconds between NetworkManager state "prepare" to state "activated" on a specific network interface. In this case the networkmanager block will show the device as disconnected (×) even though it eventually received IP and is working properly. The status will not change unless I reload i3 (or another network event happens).

As far as I understand, the networkmanager block will be triggered by dbus events generated by NetworkManager components, but I couldn't figure out in which moment the block stops processing and the "activated" state doesn't get through. Maybe it would be useful to add polling mechanism next to the existing event driven? That would also help in situations where you manually add one or more temporary static IPs to an interface besides the DHCP offered one.

Maybe the issue will not happen in most of the network environments, but additional flexibility could be of great use for sure.
In any case, thanks for the great piece of software!

**Comment by damirsamardzic** [comment]
Update:

While testing further, I see that the issue happens with wifi interfaces as well, maybe even more often. Looks like we get the state too soon and leave it like that. Does anyone else experience this issue?

My current workaround is to reload i3status-rs on click:
```
[[block]]
block = "networkmanager"
on_click = '''zsh -c "kill -SIGUSR1 $(pidof i3status-rs)"'''
```

**Comment by ammgws** [comment]
It might be easiest to start debugging this by adding a print statement (`eprintln!("{:#?}", event)`) or similar here:

https://github.com/greshake/i3status-rust/blob/2b63bcd6c77f8d30155c1d41cac92b6a83f4d67e/src/blocks/networkmanager.rs#L563-L564

Then we can at least see what events are coming in. If we find an event that we should be dealing with but seemingly aren't, we can then start tracking it further elsewhere in the code.

**Comment by damirsamardzic** [comment]
I'm sorry for late reply, here is config and log which shows events and missing IP info on eth0 (USB ethernet adapter) interface:
https://pastebin.com/BQLgFEN6

**Comment by ammgws** [comment]
Hmm, could you try `busctl monitor org.freedesktop.NetworkManager` and see what signals there are when eth0 finally gets an IP?

**Comment by damirsamardzic** [comment]
Here is the log file with busctl monitor running in parallel with i3status-rs:
https://www.dropbox.com/s/7jg30za7uehi5ga/i3status-rs_busctl.log?dl=0

I can see that eth0 received IP address (192.168.1.186) but i3status was not updated after that (still shows ×).

**Comment by ammgws** [comment]
Judging from the bar output (`"color":"#b58900","full_text":"  eth0: × "`) the widget state is being set to Warning and the ip to the default of "x", so at least that gives some context. Think it's just that we're not catching some signals that would cause the block to udpate.

**Comment by ammgws** [comment]
Closed via #991
